Finality provider crashes over duplicated finality signature submission

We've definitely seen this before; an FP is trying to send a finality signature that has been sent already while catching up.

```shell
+ fpd start
service injective.evm.v1beta1.Msg does not have cosmos.msg.v1.service proto annotation
2024-01-30T14:03:19.709308Z	info	successfully connected to a remote EOTS manager	{"address": "eotsmanager.finality-provider:15813"}
2024-01-30T14:03:19.711357Z	info	Starting FinalityProviderApp
2024-01-30T14:03:19.711585Z	info	Starting finality-provider instance	{"pk": "043416ccead8286315a204a072fd00db295470876f7ed32b84c6f293dbd38ffa"}
2024-01-30T14:03:27.179496Z	warn	Sent transaction but received failure response	{"chain_id": "bbn-dev-5", "msg_types": ["/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig", "/babylon.finality.v1.MsgAddFinalitySig"], "response": {"height": 6884, "tx_hash": "7AC7BF0371A03CC2AF02A5BABD1B21664DF7DC923ECA2B15B402AE0932BA8FE0", "codespace": "finality", "code": 1109, "data": "", "events": [{"event_type": "coin_spent", "event_attr: spender": "bbn1eh52lqv4ncxvflky6mld49lxx0j9eg05yr7xwu", "event_attr: amount": "1953ubbn"}, {"event_type": "coin_received", "event_attr: receiver": "bbn17xpfvakm2amg962yls6f84z3kell8c5l88j35y", "event_attr: amount": "1953ubbn"}, {"event_type": "transfer", "event_attr: recipient": "bbn17xpfvakm2amg962yls6f84z3kell8c5l88j35y", "event_attr: sender": "bbn1eh52lqv4ncxvflky6mld49lxx0j9eg05yr7xwu", "event_attr: amount": "1953ubbn"}, {"event_type": "message", "event_attr: sender": "bbn1eh52lqv4ncxvflky6mld49lxx0j9eg05yr7xwu"}, {"event_type": "tx", "event_attr: fee": "1953ubbn", "event_attr: fee_payer": "bbn1eh52lqv4ncxvflky6mld49lxx0j9eg05yr7xwu"}, {"event_type": "tx", "event_attr: acc_seq": "bbn1eh52lqv4ncxvflky6mld49lxx0j9eg05yr7xwu/2753"}, {"event_type": "tx", "event_attr: signature": "Sz5ZOBlTm2g0a7Q0y7EAiy/427D+WuWl/xB6o8s4IzUvfcNGFj9xOLPD140OjObBEjnuy/ok1OX3G5rWyApcNg=="}]}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x2c55cdb]

goroutine 1 [running]:
github.com/babylonchain/finality-provider/clientcontroller.(*BabylonController).SubmitBatchFinalitySigs(0xc00156c040, 0xc001209520?, {0xc002605450, 0xa, 0x45fe65?}, {0xc00270dbd0, 0x50?, 0x50?})
	/go/src/github.com/babylonchain/finality-provider/clientcontroller/babylon.go:217 +0x35b
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).SubmitBatchFinalitySignatures(0xc00260a640, {0xc002605450?, 0xa, 0xa})
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fp_instance.go:677 +0x1fc
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).FastSync(0xc00260a640, 0x3fcfe8b?, 0x1ae3)
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fastsync.go:74 +0x2c9
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).tryFastSync(0xc00260a640, 0xc002711f50)
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fp_instance.go:388 +0x3ae
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).bootstrap(0xc00260a640)
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fp_instance.go:139 +0x65
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).Start(0xc00260a640)
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fp_instance.go:102 +0x13c
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderManager).addFinalityProviderInstance(0xc002602310, 0x2e84bd501586b4f?, {0x0, 0x0})
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fp_manager.go:339 +0x245
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderManager).StartFinalityProvider(0xc002602310, 0xc000f9f4b0?, {0x0, 0x0})
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fp_manager.go:214 +0x11c
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderManager).StartAll(0xc002602310)
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/fp_manager.go:239 +0x13f
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderApp).StartHandlingAll(...)
	/go/src/github.com/babylonchain/finality-provider/finality-provider/service/app.go:218
main.start(0xc000f7f1e0)
	/go/src/github.com/babylonchain/finality-provider/finality-provider/cmd/fpd/start.go:95 +0x47a
github.com/urfave/cli.HandleAction({0x38f8940?, 0x42d36c8?}, 0x5?)
	/go/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:524 +0x50
github.com/urfave/cli.Command.Run({{0x3f41d18, 0x5}, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x3f90214, 0x1f}, {0x0, ...}, ...}, ...)
	/go/pkg/mod/github.com/urfave/cli@v1.22.14/command.go:175 +0x63e
github.com/urfave/cli.(*App).Run(0xc001482a80, {0xc000150000, 0x2, 0x2})
	/go/pkg/mod/github.com/urfave/cli@v1.22.14/app.go:277 +0xb27
main.main()
	/go/src/github.com/babylonchain/finality-provider/finality-provider/cmd/fpd/main.go:22 +0x2a5
```

```shell
babylon@genesis-validator2-788b9f8c9f-svhq4:~$ babylond query tx 7AC7BF0371A03CC2AF02A5BABD1B21664DF7DC923ECA2B15B402AE0932BA8FE0 | jq
{
  "height": "6884",
  "txhash": "7AC7BF0371A03CC2AF02A5BABD1B21664DF7DC923ECA2B15B402AE0932BA8FE0",
  "codespace": "finality",
  "code": 1109,
  "data": "",
  "raw_log": "failed to execute message; message index: 0: the finality signature has been casted before",
```

--- a/clientcontroller/babylon.go
+++ b/clientcontroller/babylon.go
@@ -214,9 +214,9 @@ func (bc *BabylonController) SubmitFinalitySig(fpPk *btcec.PublicKey, blockHeigh
 		return nil, err
 	}
 	if res == nil {
-		// NOTE: this can happen when encountering ErrDuplicatedFinalitySig
+		// NOTE: this can happen when encountering expected errors
 		// during retrying
-		return &types.TxResponse{}, nil
+		return nil, fmt.Errorf("encountered expected err: %w", Expected(err))
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil
@@ -255,9 +255,9 @@ func (bc *BabylonController) SubmitBatchFinalitySigs(fpPk *btcec.PublicKey, bloc
 		return nil, err
 	}
 	if res == nil {
-		// NOTE: this can happen when encountering ErrDuplicatedFinalitySig
+		// NOTE: this can happen when encountering expected errors
 		// during retrying
-		return &types.TxResponse{}, nil
+		return nil, fmt.Errorf("encountered expected err: %w", Expected(err))
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil

--- a/clientcontroller/babylon.go
+++ b/clientcontroller/babylon.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"time"
 
+	sdkErr "cosmossdk.io/errors"
 	"github.com/btcsuite/btcd/btcec/v2"
 	"github.com/btcsuite/btcd/btcutil"
 
@@ -29,6 +30,8 @@ import (
 )
 
 var _ ClientController = &BabylonController{}
+
+var emptyErrs = []*sdkErr.Error{}
 
 type BabylonController struct {
 	bbnClient *bbnclient.Client
@@ -92,25 +95,20 @@ func (bc *BabylonController) GetKeyAddress() sdk.AccAddress {
 	return addr
 }
 
-func (bc *BabylonController) reliablySendMsg(msg sdk.Msg) (*provider.RelayerTxResponse, error) {
-	return bc.reliablySendMsgs([]sdk.Msg{msg})
+func (bc *BabylonController) reliablySendMsg(msg sdk.Msg, expectedErrs []*sdkErr.Error, unrecoverableErrs []*sdkErr.Error) (*provider.RelayerTxResponse, error) {
+	return bc.reliablySendMsgs([]sdk.Msg{msg}, expectedErrs, unrecoverableErrs)
 }
 
-func (bc *BabylonController) reliablySendMsgs(msgs []sdk.Msg) (*provider.RelayerTxResponse, error) {
+func (bc *BabylonController) reliablySendMsgs(msgs []sdk.Msg, expectedErrs []*sdkErr.Error, unrecoverableErrs []*sdkErr.Error) (*provider.RelayerTxResponse, error) {
 	res, err := bc.bbnClient.ReliablySendMsgs(
 		context.Background(),
 		msgs,
-		expectedErrors,
-		unrecoverableErrors,
+		expectedErrs,
+		unrecoverableErrs,
 	)
 
 	if err != nil {
 		return nil, err
-	}
-
-	if res == nil {
-		// this can happen when the error is expected during retrying
-		return &provider.RelayerTxResponse{}, nil
 	}
 
 	return res, nil
@@ -144,7 +142,7 @@ func (bc *BabylonController) RegisterFinalityProvider(
 		Description: &sdkDescription,
 	}
 
-	res, err := bc.reliablySendMsg(msg)
+	res, err := bc.reliablySendMsg(msg, emptyErrs, emptyErrs)
 	if err != nil {
 		return nil, err
 	}
@@ -176,7 +174,14 @@ func (bc *BabylonController) CommitPubRandList(
 		Sig:         bip340Sig,
 	}
 
-	res, err := bc.reliablySendMsg(msg)
+	unrecoverableErrs := []*sdkErr.Error{
+		finalitytypes.ErrInvalidPubRand,
+		finalitytypes.ErrTooFewPubRand,
+		finalitytypes.ErrNoPubRandYet,
+		btcstakingtypes.ErrFpNotFound,
+	}
+
+	res, err := bc.reliablySendMsg(msg, emptyErrs, unrecoverableErrs)
 	if err != nil {
 		return nil, err
 	}
@@ -194,9 +199,24 @@ func (bc *BabylonController) SubmitFinalitySig(fpPk *btcec.PublicKey, blockHeigh
 		FinalitySig:  bbntypes.NewSchnorrEOTSSigFromModNScalar(sig),
 	}
 
-	res, err := bc.reliablySendMsg(msg)
+	expectedErrs := []*sdkErr.Error{
+		finalitytypes.ErrDuplicatedFinalitySig,
+	}
+
+	unrecoverableErrs := []*sdkErr.Error{
+		finalitytypes.ErrInvalidFinalitySig,
+		finalitytypes.ErrPubRandNotFound,
+		btcstakingtypes.ErrFpAlreadySlashed,
+	}
+
+	res, err := bc.reliablySendMsg(msg, expectedErrs, unrecoverableErrs)
 	if err != nil {
 		return nil, err
+	}
+	if res == nil {
+		// NOTE: this can happen when encountering ErrDuplicatedFinalitySig
+		// during retrying
+		return &types.TxResponse{}, nil
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil
@@ -220,9 +240,24 @@ func (bc *BabylonController) SubmitBatchFinalitySigs(fpPk *btcec.PublicKey, bloc
 		msgs = append(msgs, msg)
 	}
 
-	res, err := bc.reliablySendMsgs(msgs)
+	expectedErrs := []*sdkErr.Error{
+		finalitytypes.ErrDuplicatedFinalitySig,
+	}
+
+	unrecoverableErrs := []*sdkErr.Error{
+		finalitytypes.ErrInvalidFinalitySig,
+		finalitytypes.ErrPubRandNotFound,
+		btcstakingtypes.ErrFpAlreadySlashed,
+	}
+
+	res, err := bc.reliablySendMsgs(msgs, expectedErrs, unrecoverableErrs)
 	if err != nil {
 		return nil, err
+	}
+	if res == nil {
+		// NOTE: this can happen when encountering ErrDuplicatedFinalitySig
+		// during retrying
+		return &types.TxResponse{}, nil
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil
@@ -414,7 +449,7 @@ func (bc *BabylonController) CreateBTCDelegation(
 		DelegatorUnbondingSlashingSig: delUnbondingSlashingSig,
 	}
 
-	res, err := bc.reliablySendMsg(msg)
+	res, err := bc.reliablySendMsg(msg, emptyErrs, emptyErrs)
 	if err != nil {
 		return nil, err
 	}
@@ -428,7 +463,7 @@ func (bc *BabylonController) InsertBtcBlockHeaders(headers []bbntypes.BTCHeaderB
 		Headers: headers,
 	}
 
-	res, err := bc.reliablySendMsg(msg)
+	res, err := bc.reliablySendMsg(msg, emptyErrs, emptyErrs)
 	if err != nil {
 		return nil, err
 	}
@@ -556,7 +591,7 @@ func (bc *BabylonController) SubmitCovenantSigs(
 		SlashingUnbondingTxSigs: unbondingSlashingSigs,
 	}
 
-	res, err := bc.reliablySendMsg(msg)
+	res, err := bc.reliablySendMsg(msg, emptyErrs, emptyErrs)
 	if err != nil {
 		return nil, err
 	}

--- a/clientcontroller/babylon.go
+++ b/clientcontroller/babylon.go
@@ -100,18 +100,12 @@ func (bc *BabylonController) reliablySendMsg(msg sdk.Msg, expectedErrs []*sdkErr
 }
 
 func (bc *BabylonController) reliablySendMsgs(msgs []sdk.Msg, expectedErrs []*sdkErr.Error, unrecoverableErrs []*sdkErr.Error) (*provider.RelayerTxResponse, error) {
-	res, err := bc.bbnClient.ReliablySendMsgs(
+	return bc.bbnClient.ReliablySendMsgs(
 		context.Background(),
 		msgs,
 		expectedErrs,
 		unrecoverableErrs,
 	)
-
-	if err != nil {
-		return nil, err
-	}
-
-	return res, nil
 }
 
 // RegisterFinalityProvider registers a finality provider via a MsgCreateFinalityProvider to Babylon

--- a/clientcontroller/babylon.go
+++ b/clientcontroller/babylon.go
@@ -210,7 +210,7 @@ func (bc *BabylonController) SubmitFinalitySig(fpPk *btcec.PublicKey, blockHeigh
 	if res == nil {
 		// NOTE: this can happen when encountering expected errors
 		// during retrying
-		return nil, Expected(fmt.Errorf("expected error"))
+		return nil, Expected(fmt.Errorf("expected error: %w", finalitytypes.ErrDuplicatedFinalitySig))
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil
@@ -251,7 +251,7 @@ func (bc *BabylonController) SubmitBatchFinalitySigs(fpPk *btcec.PublicKey, bloc
 	if res == nil {
 		// NOTE: this can happen when encountering expected errors
 		// during retrying
-		return nil, Expected(fmt.Errorf("expected error"))
+		return nil, Expected(fmt.Errorf("expected error: %w", finalitytypes.ErrDuplicatedFinalitySig))
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil

--- a/clientcontroller/babylon.go
+++ b/clientcontroller/babylon.go
@@ -216,7 +216,7 @@ func (bc *BabylonController) SubmitFinalitySig(fpPk *btcec.PublicKey, blockHeigh
 	if res == nil {
 		// NOTE: this can happen when encountering expected errors
 		// during retrying
-		return nil, fmt.Errorf("encountered expected err: %w", Expected(err))
+		return nil, Expected(fmt.Errorf("expected error"))
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil
@@ -257,7 +257,7 @@ func (bc *BabylonController) SubmitBatchFinalitySigs(fpPk *btcec.PublicKey, bloc
 	if res == nil {
 		// NOTE: this can happen when encountering expected errors
 		// during retrying
-		return nil, fmt.Errorf("encountered expected err: %w", Expected(err))
+		return nil, Expected(fmt.Errorf("expected error"))
 	}
 
 	return &types.TxResponse{TxHash: res.TxHash, Events: res.Events}, nil

--- a/clientcontroller/babylon.go
+++ b/clientcontroller/babylon.go
@@ -97,12 +97,23 @@ func (bc *BabylonController) reliablySendMsg(msg sdk.Msg) (*provider.RelayerTxRe
 }
 
 func (bc *BabylonController) reliablySendMsgs(msgs []sdk.Msg) (*provider.RelayerTxResponse, error) {
-	return bc.bbnClient.ReliablySendMsgs(
+	res, err := bc.bbnClient.ReliablySendMsgs(
 		context.Background(),
 		msgs,
 		expectedErrors,
 		unrecoverableErrors,
 	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if res == nil {
+		// this can happen when the error is expected during retrying
+		return &provider.RelayerTxResponse{}, nil
+	}
+
+	return res, nil
 }
 
 // RegisterFinalityProvider registers a finality provider via a MsgCreateFinalityProvider to Babylon

--- a/clientcontroller/retry_utils.go
+++ b/clientcontroller/retry_utils.go
@@ -13,8 +13,6 @@ import (
 var unrecoverableErrors = []*sdkErr.Error{
 	finalitytypes.ErrBlockNotFound,
 	finalitytypes.ErrInvalidFinalitySig,
-	finalitytypes.ErrHeightTooHigh,
-	finalitytypes.ErrInvalidPubRand,
 	finalitytypes.ErrNoPubRandYet,
 	finalitytypes.ErrPubRandNotFound,
 	finalitytypes.ErrTooFewPubRand,
@@ -24,23 +22,6 @@ var unrecoverableErrors = []*sdkErr.Error{
 // IsUnrecoverable returns true when the error is in the unrecoverableErrors list
 func IsUnrecoverable(err error) bool {
 	for _, e := range unrecoverableErrors {
-		if errors.Is(err, e) {
-			return true
-		}
-	}
-
-	return false
-}
-
-var expectedErrors = []*sdkErr.Error{
-	// if due to some low-level reason (e.g., network), we submit duplicated finality sig,
-	// we should just ignore the error
-	finalitytypes.ErrDuplicatedFinalitySig,
-}
-
-// IsExpected returns true when the error is in the expectedErrors list
-func IsExpected(err error) bool {
-	for _, e := range expectedErrors {
 		if errors.Is(err, e) {
 			return true
 		}

--- a/clientcontroller/retry_utils.go
+++ b/clientcontroller/retry_utils.go
@@ -29,3 +29,17 @@ func IsUnrecoverable(err error) bool {
 
 	return false
 }
+
+type ExpectedError struct {
+	error
+}
+
+// Expected wraps an error in ExpectedError struct
+func Expected(err error) error {
+	return ExpectedError{err}
+}
+
+// IsExpected checks if error is an instance of ExpectedError
+func IsExpected(err error) bool {
+	return errors.Is(err, ExpectedError{})
+}

--- a/clientcontroller/retry_utils.go
+++ b/clientcontroller/retry_utils.go
@@ -34,6 +34,23 @@ type ExpectedError struct {
 	error
 }
 
+func (e ExpectedError) Error() string {
+	if e.error == nil {
+		return "expected error"
+	}
+	return e.error.Error()
+}
+
+func (e ExpectedError) Unwrap() error {
+	return e.error
+}
+
+// Is adds support for errors.Is usage on isExpected
+func (ExpectedError) Is(err error) bool {
+	_, isExpected := err.(ExpectedError)
+	return isExpected
+}
+
 // Expected wraps an error in ExpectedError struct
 func Expected(err error) error {
 	return ExpectedError{err}

--- a/clientcontroller/retry_utils_test.go
+++ b/clientcontroller/retry_utils_test.go
@@ -1,0 +1,16 @@
+package clientcontroller
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/babylonchain/babylon/x/finality/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestExpectedErr(t *testing.T) {
+	expectedErr := Expected(types.ErrDuplicatedFinalitySig)
+	require.True(t, IsExpected(expectedErr))
+	wrappedErr := fmt.Errorf("expected: %w", expectedErr)
+	require.True(t, IsExpected(wrappedErr))
+}

--- a/finality-provider/service/fp_instance.go
+++ b/finality-provider/service/fp_instance.go
@@ -513,6 +513,11 @@ func (fp *FinalityProviderInstance) retrySubmitFinalitySignatureUntilBlockFinali
 			if clientcontroller.IsUnrecoverable(err) {
 				return nil, err
 			}
+
+			if clientcontroller.IsExpected(err) {
+				return nil, nil
+			}
+
 			fp.logger.Debug(
 				"failed to submit finality signature to the consumer chain",
 				zap.String("pk", fp.GetBtcPkHex()),

--- a/finality-provider/service/fp_instance.go
+++ b/finality-provider/service/fp_instance.go
@@ -137,7 +137,7 @@ func (fp *FinalityProviderInstance) bootstrap() (uint64, error) {
 
 	if fp.checkLagging(latestBlock) {
 		_, err := fp.tryFastSync(latestBlock)
-		if err != nil {
+		if err != nil && !clientcontroller.IsExpected(err) {
 			return 0, err
 		}
 	}

--- a/finality-provider/service/fp_instance.go
+++ b/finality-provider/service/fp_instance.go
@@ -510,13 +510,6 @@ func (fp *FinalityProviderInstance) retrySubmitFinalitySignatureUntilBlockFinali
 		// error will be returned if max retries have been reached
 		res, err := fp.SubmitFinalitySignature(targetBlock)
 		if err != nil {
-			if clientcontroller.IsUnrecoverable(err) {
-				return nil, err
-			}
-
-			if clientcontroller.IsExpected(err) {
-				return nil, nil
-			}
 
 			fp.logger.Debug(
 				"failed to submit finality signature to the consumer chain",
@@ -525,6 +518,14 @@ func (fp *FinalityProviderInstance) retrySubmitFinalitySignatureUntilBlockFinali
 				zap.Uint64("target_block_height", targetBlock.Height),
 				zap.Error(err),
 			)
+
+			if clientcontroller.IsUnrecoverable(err) {
+				return nil, err
+			}
+
+			if clientcontroller.IsExpected(err) {
+				return nil, nil
+			}
 
 			failedCycles += 1
 			if failedCycles > uint32(fp.cfg.MaxSubmissionRetries) {

--- a/itest/e2e_test.go
+++ b/itest/e2e_test.go
@@ -191,3 +191,65 @@ func TestFastSync(t *testing.T) {
 	require.NoError(t, err)
 	require.True(t, currentHeight < finalizedHeight+uint64(n))
 }
+
+// TestFastSync_DuplicateVotes covers a special case when the finality signature
+// has inconsistent view of last voted height with the Babylon node and during
+// fast-sync it submits a batch of finality sigs, one of which is rejected due
+// to duplicate error
+// this test covers this case by starting 3 finality providers, 2 of which
+// are stopped after gaining voting power to simulate the case where no blocks
+// are finalized. Then we let one of the finality providers "forget" the last
+// voted height and restart all the finality providers, expecting it them to
+// catch up and finalize new blocks
+func TestFastSync_DuplicateVotes(t *testing.T) {
+	tm, fpInsList := StartManagerWithFinalityProvider(t, 3)
+	defer tm.Stop(t)
+
+	fpIns1 := fpInsList[0]
+	fpIns2 := fpInsList[1]
+	fpIns3 := fpInsList[2]
+
+	// check the public randomness is committed
+	tm.WaitForFpPubRandCommitted(t, fpIns1)
+	tm.WaitForFpPubRandCommitted(t, fpIns2)
+	tm.WaitForFpPubRandCommitted(t, fpIns3)
+
+	// send 3 BTC delegations to empower the three finality providers
+	_ = tm.InsertBTCDelegation(t, []*btcec.PublicKey{fpIns1.MustGetBtcPk()}, stakingTime, stakingAmount)
+	_ = tm.InsertBTCDelegation(t, []*btcec.PublicKey{fpIns2.MustGetBtcPk()}, stakingTime, stakingAmount)
+	_ = tm.InsertBTCDelegation(t, []*btcec.PublicKey{fpIns3.MustGetBtcPk()}, stakingTime, stakingAmount)
+
+	// check the BTC delegations are pending
+	dels := tm.WaitForNPendingDels(t, 3)
+
+	// send covenant sigs to each delegation
+	tm.InsertCovenantSigForDelegation(t, dels[0])
+	tm.InsertCovenantSigForDelegation(t, dels[1])
+	tm.InsertCovenantSigForDelegation(t, dels[2])
+
+	// check the BTC delegations are active
+	_ = tm.WaitForNActiveDels(t, 3)
+
+	// stop 2 of the finality providers so that no blocks will be finalized
+	err := fpIns2.Stop()
+	require.NoError(t, err)
+	err = fpIns3.Stop()
+	require.NoError(t, err)
+
+	// make sure fp1 has cast a finality vote
+	// and then make it "forget" the last voted height
+	lastVotedHeight := tm.WaitForFpVoteCast(t, fpIns1)
+	fpIns1.MustUpdateStateAfterFinalitySigSubmission(lastVotedHeight - 1)
+
+	// stop fp1, restarts all the fps after 3 blocks for them to catch up
+	// and expect a block will be finalized
+	n := 3
+	tm.FpConfig.FastSyncGap = uint64(n)
+	tm.StopAndRestartFpAfterNBlocks(t, n, fpIns1)
+	err = fpIns2.Start()
+	require.NoError(t, err)
+	err = fpIns3.Start()
+	require.NoError(t, err)
+	finalizedBlocks := tm.WaitForNFinalizedBlocks(t, 1)
+	require.GreaterOrEqual(t, finalizedBlocks[0].Height, lastVotedHeight)
+}

